plugins/sinks/shield/sink.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	return nil
+}
+
+func (s *Sink) Sink(ctx context.Context, batch []models.Record) error {
+	for _, record := range batch {
+		metadata := record.Data()
+		s.logger.Info("sinking record to shield", "record", metadata.GetUrn())
+
+		shieldPayload, err := s.buildShieldPayload(metadata)
+		if err != nil {
+			return errors.Wrap(err, "failed to build shield payload")
+		}
+		if err = s.send(ctx, shieldPayload); err != nil {
+			return errors.Wrap(err, "error sending data")
+		}
+
+		s.logger.Info("successfully sinked record to shield", "record", metadata.Name)
+	}
+
+	return nil
+}
+
+func (s *Sink) Close() (err error) {
+	return
+	//TODO: Connection closes even when some records are unpiblished
+	//TODO: return s.client.Close()
+}
+
+func (s *Sink) send(ctx context.Context, record RequestPayload) error {
+	for hdrKey, hdrVal := range s.config.Headers {
+		hdrVals := strings.Split(hdrVal, ",")
+		for _, val := range hdrVals {
+			val = strings.TrimSpace(val)
+			md := metadata.New(map[string]string{hdrKey: val})
+			ctx = metadata.NewOutgoingContext(ctx, md)
+		}
+	}
+
+	metadataBytes, err := json.Marshal(record.Metadata)
+	if err != nil {
+		return err
+	}
+
+	var requestBody sh.UserRequestBody
+	requestBody.Name = record.Name
+	requestBody.Email = record.Email
+	requestBody.Metadata = new(structpb.Struct)
+
+	err = json.Unmarshal(metadataBytes, requestBody.Metadata)
+	if err != nil {
+		return err
+	}
+
+	_, err = s.client.UpdateUser(ctx, &sh.UpdateUserRequest{
+		Id:   requestBody.Email,
+		Body: &requestBody,
+	})
+	if err == nil {
+		return nil
+	}
+
+	if e, ok := status.FromError(err); ok {
+		err = fmt.Errorf("shield returns code %d: %v", e.Code(), e.Message())
+		switch e.Code() {
+		case codes.Unavailable:
+			return plugins.NewRetryError(err)
+		default:
+			return err
+		}
+	} else {
+		err = fmt.Errorf("not able to parse error returned %v", err)
+	}
+
+	return err
+}
+
+func (s *Sink) buildShieldPayload(resource *assetsv1beta2.Asset) (RequestPayload, error) {
+	data := resource.GetData()
+
+	mapdata, err := s.buildShieldData(data)
+	if err != nil {
+		return RequestPayload{}, err
+	}
+
+	name, ok := mapdata["full_name"].(string)
+	if !ok {
+		return RequestPayload{}, errors.New(fmt.Sprintf("unexpected type %T for name, must be a string", mapdata["full_name"]))
+	}
+
+	email, ok := mapdata["email"].(string)
+	if !ok {
+		return RequestPayload{}, errors.New(fmt.Sprintf("unexpected type %T for email, must be a string", mapdata["email"]))
+	}
+
+	metadata, ok := mapdata["attributes"].(map[string]interface{})
+	if !ok {
+		return RequestPayload{}, errors.New(fmt.Sprintf("unexpected type %T for attributes, must be a map[string]interface{}", mapdata["attributes"]))
+	}
+
+	record := RequestPayload{
+		Name:     name,
+		Email:    email,
+		Metadata: metadata,
+	}
+
+	return record, nil
+}
+
+func (s *Sink) buildShieldData(anyData *anypb.Any) (map[string]interface{}, error) {
+	var mapData map[string]interface{}
```

Comment:
Instead of `mapData` we can directly unmarshal it to `v1beta2.User`

```
import "github.com/odpf/meteor/models/odpf/assets/v1beta2"

var user v1beta2.User
err := anyData.UnmarshalTo(&user)
```